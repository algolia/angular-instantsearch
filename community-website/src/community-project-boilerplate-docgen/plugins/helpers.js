const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
