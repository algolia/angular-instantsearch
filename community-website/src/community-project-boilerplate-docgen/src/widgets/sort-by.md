---
title: Widgets - Sort By
layout: widget.pug
---

# Sort by

## Description

The SortBy component displays a list of indexes allowing a user to change the hits are sorting.

## Options

| Attributes  | Type                              | Description
| -           | -                                 | -
| `items`     | `{name: string, label: string}[]` | The list of indexes to search in

## Code example

```html
<ng-ais-sort-by
  [items]="[
    { name: 'instant_search', label: 'Most relevant' },
    { name: 'instant_search_price_asc', label: 'Lowest price' },
    { name: 'instant_search_price_desc', label: 'Highest price' }
  ]"
>
</ng-ais-sort-by>
```
