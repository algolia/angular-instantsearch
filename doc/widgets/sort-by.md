##### `<ng-ais-sort-by />`

The SortBy component displays a list of indexes allowing a user to change the hits are sorting.

| Attributes  | Type                              | Description
| -           | -                                 | -
| `items`     | `{name: string, label: string}[]` | The list of indexes to search in
