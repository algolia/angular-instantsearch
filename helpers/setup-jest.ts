import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
