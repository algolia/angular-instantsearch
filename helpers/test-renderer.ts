import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
