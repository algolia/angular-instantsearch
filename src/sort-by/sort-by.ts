import { Component, Input, Inject, forwardRef } from "@angular/core";

import { connectSortBySelector } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";

@Component({
  selector: "ng-ais-sort-by",
  template: `
    <div [class]="cx()">
      <select
        [class]="cx('select')"
        (change)="state.refine($event.target.value)"
      >
        <option
          [class]="cx('option')"
          *ngFor="let item of state.options"
          [value]="item.value"
          [selected]="item.value === state.currentRefinement"
        >
          {{item.label}}
        </option>
      </select>
    </div>
  `
})
export class NgAisSortBy extends BaseWidget {
  @Input()
  public items: {
    name: string;
    label: string;
  }[];

  public state: {
    currentRefinement: string | null;
    options: {}[];
    refine: Function;
  } = {
    currentRefinement: null,
    options: [],
    refine: noop
  };

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super("SortBy");
  }

  public ngOnInit() {
    this.createWidget(connectSortBySelector, { indices: this.items });
    super.ngOnInit();
  }
}
