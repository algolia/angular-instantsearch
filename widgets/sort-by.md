##### `<ng-ais-sort-by />`

The SortBy component displays a list of indexes allowing a user to change the hits are sorting.

| Attributes  | Type                              | Description
| -           | -                                 | -
| `indices`   | `{name: string, label: string}[]` | The list of indexes to search in
| `header?`   | `string`                          | Displays text before widget
| `footer?`   | `string`                          | Displays text after widget
