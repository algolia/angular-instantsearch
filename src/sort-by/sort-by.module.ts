import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisSortBy } from "./sort-by";

@NgModule({
  declarations: [NgAisSortBy],
  entryComponents: [NgAisSortBy],
  exports: [NgAisSortBy],
  imports: [CommonModule]
})
export class NgAisSortByModule {}
