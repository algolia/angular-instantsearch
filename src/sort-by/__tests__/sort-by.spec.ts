import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisSortBy } from '../sort-by';

const defaultState = {
  currentRefinement: 'foo',
  options: [{ label: 'foo', value: 'foo' }, { label: 'bar', value: 'bar' }],
  refine: jest.fn(),
};

const render = createRenderer({
  defaultState,
  template: '<ais-sort-by></ais-sort-by>',
  TestedWidget: NgAisSortBy,
});

describe('SortBy', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should call refine on select change', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const select = fixture.debugElement.nativeElement.querySelector('select');
    select.value = 'bar';
    select.dispatchEvent(new Event('change'));

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith('bar');
  });
});
