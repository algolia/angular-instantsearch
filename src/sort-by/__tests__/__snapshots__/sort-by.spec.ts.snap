// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortBy renders markup with state 1`] = `
<ng-ais-sort-by
  cx={[Function Function]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-SortBy"
  >
    
        
    <ng-ais-header
      ng-reflect-class-name="ais-SortBy-header ais-header"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-SortBy-body ais-body"
    >
      
          
      <select
        class="ais-SortBy-select"
      >
        
            
        
        <option
          class="ais-SortBy-option"
          value="foo"
        >
          
              foo
            
        </option>
        <option
          class="ais-SortBy-option"
          value="bar"
        >
          
              bar
            
        </option>
        
          
      </select>
      
        
    </div>
    
  
        
    <ng-ais-footer
      ng-reflect-class-name="ais-SortBy-footer ais-footer"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-sort-by>
`;

exports[`SortBy renders markup without state 1`] = `
<ng-ais-sort-by
  cx={[Function Function]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-SortBy"
  >
    
        
    <ng-ais-header
      ng-reflect-class-name="ais-SortBy-header ais-header"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-SortBy-body ais-body"
    >
      
          
      <select
        class="ais-SortBy-select"
      >
        
            
        
        
          
      </select>
      
        
    </div>
    
  
        
    <ng-ais-footer
      ng-reflect-class-name="ais-SortBy-footer ais-footer"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-sort-by>
`;
